[ADD] procurement_sale_forecast & procurement_purchase_forecast

agaldona opened the pull request.

Commit:
[ADD] procurement_sale_forecast

agaldona commented:
Aqui está el módulo de @oihane pero está muy retocado. He intentado hacer lo que hablamos con @anajuaristi la última reunión. 
Falta la parte de compras, que en breve podré subir en otro módulo.

oihane commented:
Pues me voy a adelantar... y decirte que.... vas a tener que pasarlo a la nueva API :stuck_out_tongue_closed_eyes:

agaldona commented:
@oihane el que?

Commit:
[IMP] pasar a la nueva api el fichero de procurement order

oihane commented on procurement_sale_forecast/models/procurement.py:
```diff
@@ -0,0 +1,29 @@
+from openerp.osv import orm, fields
+
+
+class ProcurementOrder(orm.Model):
```
Esto hay que pasarlo a la nueva API

agaldona replied:
yasta! :wink: me he adelantado... se me había colado

oihane commented on procurement_sale_forecast/security/ir.model.access.csv:
```diff
@@ -0,0 +1,5 @@
+access_procurement_sale_forecast_sale_manager,procurement.sale.forecast manager,model_procurement_sale_forecast,base.group_sale_manager,1,1,1,1
+access_procurement_sale_forecast_line,procurement.sale.forecast.line user,model_procurement_sale_forecast_line,base.group_user,1,0,0,0
+access_procurement_sale_forecast_sale_line_manager,procurement.sale.forecast.line manager,model_procurement_sale_forecast_line,base.group_sale_manager,1,1,1,1
+access_load_sales_on_forecast_manager,load.sales.on.forecast manager,model_load_sales_on_forecast,base.group_sale_manager,1,1,1,1
```
Hay que poner un salto de línea, aquí

Commit:
[FIX] nueva línea al final en el fichero de permisos

Commit:
[IMP] Volver a pasar Flake8

Commit:
[ADD] procurement_purchase_forecast

agaldona commented:
`procurement_purchase_forecast` He subido aquí la parte de compras, que si subo en dos PR-s distintos me da error el travis por dependencias no resueltas

anajuaristi commented:
Probado módulo sale_forecast. 

Cosas que he visto.
En previsión:
1. En el wizard de importar, tanto de pedidos de venta como de compra, el campo partner es obligatorio. No debería serlo ya que en caso de que se quiera hacer una previsión para un partner determinado, esa opción ya está en la cabecera. 
Lo que sí debería hacer es que si se pone un partner en cabecera, por defecto el wizard tendría que coger ese. Pero es necesario permitir que se puedan importar líneas de pedidos independientemente del partner, de hecho, creo que partner debería ser un campo asociado a la línea y no a la cabecera.
Igualmente añadiría la opción de poder asignar una línea de previsión a un comercial, y poder importar ventas de un determinado comercial, añadiendo el filtro en el wizard 
2. En el wizard falta el campo "factor". Es posible que quieran subir un % determinado, o bajar, las previsiones de un periodo al siguiente. Por defecto sería un 1, pero si meten un valor multiplicaría por dicho valor las cantidades de las líneas importadas. 
3. El botón de crear abastecimientos, en cuanto se pulse una vez, debería quedar desactivado. Actualmente cuando pulsas varias veces, crea un abastecimiento cada vez 
4. Como mejora, añadiría un shorcut en cabecera a abastecimientos creados desde la previsión.

---

En líneas de previsión
1. Los group by dan traceback (el de fechas, el de producto... no he probado el resto)

```
  File "/opt/odoo/server-v8/openerp/http.py", line 387, in response_wrap
    response = f(*args, **kw)
  File "/opt/odoo/server-v8/addons/web/controllers/main.py", line 949, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/opt/odoo/server-v8/addons/web/controllers/main.py", line 941, in _call_kw
    return getattr(request.registry.get(model), method)(request.cr, request.uid, *args, **kwargs)
  File "/opt/odoo/server-v8/openerp/api.py", line 237, in wrapper
    return old_api(self, *args, **kwargs)
  File "/opt/odoo/server-v8/openerp/models.py", line 2094, in read_group
    assert groupby_def and groupby_def._classic_write, "Fields in 'groupby' must be regular database-persisted fields (no function or related fields), or function fields with store=True"
AssertionError: Fields in 'groupby' must be regular database-persisted fields (no function or related fields), or function fields with store=True

```

Por lo demás, funcionalmente OK, por mi parte RTM :+1:

anajuaristi commented:
Por último, purchase_forecast Ok, RTM :+1:

pedrobaeza commented on procurement_purchase_forecast/__openerp__.py:
```diff
@@ -0,0 +1,38 @@
+    "website": "http://www.odoomrp.com",
+    "complexity": "normal",
+    "description": """
+        - New wizard on purchase orders and forecast to load purchase lines on
```
Poner descripción en archivo README.rst

pedrobaeza commented on procurement_purchase_forecast/models/procurement_sale_forecast.py:
```diff
@@ -0,0 +1,47 @@
+
+class SaleForecastProcurement(models.Model):
+
+    _inherit = 'sale.forecast.procurement'
```
Por qué el nombre del modelo al revés que el del módulo?

pedrobaeza commented on procurement_purchase_forecast/security/ir.model.access.csv:
```diff
@@ -0,0 +1 @@
+access_load_purchases_on_forecast_manager,load.purchases.on.forecast manager,model_load_purchases_on_forecast,purchase.group_purchase_manager,1,1,1,1
```
No necesitas reglas de acceso para un TransientModel.

pedrobaeza commented on procurement_purchase_forecast/models/purchase_order.py:
```diff
@@ -0,0 +1,46 @@
+    _inherit = 'purchase.order'
+
+    @api.multi
+    def load_lines_on_forecast(self):
```
El nombre debería ser _load_lines_to_forecast_

pedrobaeza commented on procurement_purchase_forecast/models/purchase_order.py:
```diff
@@ -0,0 +1,46 @@
+    @api.multi
+    def load_lines_on_forecast(self):
+        self.ensure_one()
+        load_wiz_obj = self.env['load.purchases.on.forecast']
```
El nombre del TransientModel no es el más acertado. Muchas veces, no hay que poner el nombre entero, sólo algo que sea bastante significativo y que siga las otras convenciones. En este caso, creo que valdría con _purchase.forecast.load_

pedrobaeza commented on procurement_sale_forecast/models/sale.py:
```diff
@@ -0,0 +1,161 @@
+    @api.multi
+    def load_sales_on_forecast(self):
+        self.ensure_one()
+        load_wiz_obj = self.env['load.sales.on.forecast']
```
Ídem que en compras

agaldona commented:
En respuesta a @anajuaristi 
1. En el wizard de importar, tanto de pedidos de venta como de compra, el campo partner es obligatorio. No debería serlo ya que en caso de que se quiera hacer una previsión para un partner determinado, esa opción ya está en la cabecera. 
   Lo que sí debería hacer es que si se pone un partner en cabecera, por defecto el wizard tendría que coger ese. Pero es necesario permitir que se puedan importar líneas de pedidos independientemente del partner, de hecho, creo que partner debería ser un campo asociado a la línea y no a la cabecera.
   Igualmente añadiría la opción de poder asignar una línea de previsión a un comercial, y poder importar ventas de un determinado comercial, añadiendo el filtro en el wizard 

-- Si meto el partner en las líneas quitando en la cabecera, puedo tener varías líneas de un mismo producto en una previsión. Creo que el objetivo era tener una línea de producto. Por eso, si existía ya el producto, no se cargaba de nuevo.
Por otra parte, por el mismo tema tenemos el problema de que solo se cargaría parte de la previsión, ya que al cargar la primera coincidencia de ese producto para un partner, no se cargarían las demás coincidencias aunque el asistente los encuentre. Porque ya existe la línea para ese producto.
Creo que hay que darle una vuelta a esto.
1. En el wizard falta el campo "factor". Es posible que quieran subir un % determinado, o bajar, las previsiones de un periodo al siguiente. Por defecto sería un 1, pero si meten un valor multiplicaría por dicho valor las cantidades de las líneas importadas.

-- Ok, sin problema.
1. El botón de crear abastecimientos, en cuanto se pulse una vez, debería quedar desactivado. Actualmente cuando pulsas varias veces, crea un abastecimiento cada vez 

-- Ok, tendría que cargar el abastecimiento generado en la línea de alguna forma, para que se sepa que se ha creado. Y depender de eso la funcionalidad del botón.
1. Como mejora, añadiría un shorcut en cabecera a abastecimientos creados desde la previsión.

-- No estoy segura de que lo pueda hacer, de momento, no hay ningún tipo de relación entre los abastecimientos que se han creado y la previsión. Igual cuando meta el abastecimiento en la línea puedo filtrarlo de alguna forma.

Commit:
[FIX] Varias modificaciones funcionales + técnicas
- Cambio de nombre de objetos y asistentes
- Añadir la descripción en README.rst
- Quitar reglas de acceso a los asistentes
- Añadir shortcut a abastecimientos
- Pasar el partner a la línea, y cargar independientemente de si se ha especificado partner o no.
- No permitir crear más de un abastecimiento por línea.
- Añadir Factor en el asistente de carga

Commit:
[FIX] No permitir crear más de un abastecimiento por cada línea

agaldona commented:
[FIX] Varias modificaciones funcionales + técnicas:: 
- Cambio de nombre de objetos y asistentes
- Añadir la descripción en README.rst
- Quitar reglas de acceso a los asistentes
- Añadir shortcut a abastecimientos
- Pasar el partner a la línea, y cargar independientemente de si se ha especificado partner o no.
- No permitir crear más de un abastecimiento por línea.
- Añadir Factor en el asistente de carga

Creo que están todos los cambios solicitados.

pedrobaeza commented on procurement_purchase_forecast/README.rst:
```diff
@@ -0,0 +1,4 @@
+MPS - Purchase procurement
+==========================
+- New wizard on purchase orders and forecast to load purchase lines on
+forecast.
```
No le pongas guión, y si se lo pones, alínea la siguiente línea a dos espacios.

Commit:
[IMP] quitar guion en README.rst

pedrobaeza commented on procurement_purchase_forecast/models/procurement_sale_forecast.py:
```diff
@@ -0,0 +1,45 @@
+    _inherit = 'procurement.sale.forecast'
+
+    @api.multi
+    def load_purchases_to_forecast(self):
```
Por qué no utilizar una acción para esto. Los datos que traspasas son vitales? El forecast_id no lo necesitas, ya que lo tienes via self.env.context['active_id'], y entiendo que el date_from y date_to son los de por defecto, pero los va a cambiar el usuario seguro, ya que lo que querrá es traer las previsiones para un periodo anterior.

agaldona replied:
El asistente se lanza tanto en la ficha de compras como en la de la previsión. Por eso, hago está función para cargar.

pedrobaeza replied:
No importa. Tú puedes chequear vía self.env.context['active_model'] de dónde procede.

agaldona replied:
Ups, acabo de darme cuenta, de que aparte de lo que dices, el asistente no está bien funcionalmente.

agaldona replied:
Vuelvo a ponerme con esto.
Ahora tengo una duda, los datos obligatorios, unicamente son el forecast_id, y las fechas. Lo de las fechas, con tal de poner requerido me valdría, pero en el caso de forecast, si vengo de una previsión lo cargo yo, y si no, lo debo de poner como obligatorio.
¿Como podría hacer eso, sin utilizar las funciones de llamada del asistente? ¿Con la nueva API se pueden cargar datos por contexto en default?

pedrobaeza replied:
Sí, se puede poner en el contexto de la acción correspondiente 'default_forecast_id': active_id.

pedrobaeza commented on procurement_purchase_forecast/models/purchase_order.py:
```diff
@@ -0,0 +1,45 @@
+
+    @api.multi
+    def load_lines_to_forecast(self):
+        self.ensure_one()
```
Ídem

pedrobaeza commented on procurement_sale_forecast/README.rst:
```diff
@@ -0,0 +1,6 @@
+MPS - Sale forecast
+===================
+This module allows to create a sale forecast
```
Línea entre medias para que el parseador reconozca la lista

pedrobaeza commented on procurement_sale_forecast/__openerp__.py:
```diff
@@ -0,0 +1,44 @@
+    ],
+    "category": "MPS",
+    "website": "http://www.odoomrp.com",
+    "complexity": "normal",
```
La clave complexity ya no se usa

Commit:
[IMP] Añadir contribuidores + salto de línea en README.rst + quitar complexity en __openerp__.py

agaldona commented:
He subido algunos de los cambios, lo del asistente lo miro con mas tranquilidad, ya que funcionalmente también, lo tengo que revisar.

Commit:
[IMP] Cambio de wizards

pedrobaeza commented:
Hay error de Travis

Commit:
[IMP] funcionalidad nueva importar ventas y compras por meses

Commit:
[FIX] quitar prints

Commit:
[FIX] añadir cabecera en ir.model.access.csv

agaldona commented:
Ha costado lo suyo, pero ya al menos el travis no da error.

He añadido un nuevo cambio funcional solicitado por Ana.
Consiste en sacar la media por mes de un producto vendido/comprado. Y hacer una imputación de esa cantidad por cada mes en el intervalo de fechas de la previsión. Para sacar la media, las fechas a tener en cuenta son las del asistente.

pedrobaeza commented on procurement_purchase_forecast/__openerp__.py:
```diff
@@ -0,0 +1,41 @@
+        "Ana Juaristi <anajuaristi@avanzosc.es>",
+    ],
+    "website": "http://www.odoomrp.com",
+    "description": """
```
Elimina esta clave.

pedrobaeza commented on procurement_sale_forecast/models/procurement.py:
```diff
@@ -0,0 +1,26 @@
+class ProcurementOrder(models.Model):
+    _inherit = 'procurement.order'
+
+    partner_id = fields.Many2one('res.partner', string="Partner",
```
ponle store=True

Commit:
[IMP] añadir store=True en partner_id de procurement + quitar descripción de __openerp__.py

agaldona commented:
- En procurement, añadir store=True al campo partner_id
- **openerp**.py quitar descripción
